internal/models/schema-recipe.go

```diff
@@ -911,26 +896,45 @@ func (n *NutritionSchema) UnmarshalJSON(data []byte) error {
 				}
 			}
 		}
+	default:
+		slog.Warn("Could not parse nutrition schema", "schema", v, "type", x)
+	}
+	return nil
+}
 
-		if val, ok := x["sodiumContent"].(string); ok {
-			n.Sodium = val
-		}
-
-		if val, ok := x["sugarContent"].(string); ok {
-			n.Sugar = val
-		}
-
-		if val, ok := x["transFatContent"].(string); ok {
-			n.TransFat = val
+// EnsureNutritionUnitForString checks for existing unit suffix for nutritional property.
+// If not already present, add unit to the input string.
+func EnsureNutritionUnitForString(nutritionString string, nutritionProperty string) string {
+	switch nutritionProperty {
+	case "calories":
+		if nutritionString == "" {
+			return strings.TrimSpace(nutritionString)
+		} else {
+			nutritionString = strings.TrimSuffix(nutritionString, "calories")
+			nutritionString = strings.TrimSuffix(nutritionString, "kcal")
+			return strings.TrimSpace(nutritionString) + " kcal"
 		}
-
-		if val, ok := x["unsaturatedFatContent"].(string); ok {
-			n.UnsaturatedFat = val
+	case "carbohydrateContent", "fatContent", "fiberContent", "proteinContent", "saturatedFatContent", "sugarContent", "transFatContent", "unsaturatedFatContent":
+		if nutritionString == "" {
+			return strings.TrimSpace(nutritionString)
+		} else {
+			nutritionString = strings.TrimSuffix(nutritionString, "gram")
+			nutritionString = strings.TrimSuffix(nutritionString, "grams")
+			nutritionString = strings.TrimSuffix(nutritionString, "g")
```

Comment:
That is more efficient for sure, and also handles strings where the unit is not as a suffix, e.g. "23 grams sugar".

As I can see from the other test cases there can be quite a bit of variation how people have written their LD+JSON objects, so we for sure need to handle the unit being in the middle of the strings, as your suggestion does.

However, I think it would be more prudent to check for the specific words (g, gram, grams) as what will otherwise happen in case someone has written their JSON object as "23 sugar grams" or "1800 sodium milligrams ".
Now I can't imagine anyone doing that, but then again, I could not imagine anyone writing "1 grams saturated fat" either since the property name already states that. 😋

Also, at least one of the test cases fails (https://entertainingwithbeth.com/orange-pecan-crumb-muffins/), where I noticed their JSON object `"sodiumContent": "68 grams sodium"` instead of milligrams, even though the actual web page says 68 mg. 🤦‍♂️
With your proposed change this would incorrectly split on the **m** inside _grams_, becoming _**"68 gra mg"**_. This seems like (or at least I hope so) an edge case, which is also a typo from the recipe author's side, so maybe we don't even want to handle this on our end. if we e.g. replace specific words, like _mg, milligrams_ and _milligram_ in this case, the output would instead become _**"68 gram mg"**_. Not much of a difference compared to  _**"68 gra mg"**_, but still a bit clearer that something went wrong with the scraping when viewing the recipe afterwards.

What is your opinion on how to handle the above?

PS. When running all the test cases, not just the ones in the schema-recipe_test.go, I also got very aware that all test cases needs to be validated regarding the nutritional information, as my implementation now always converts to the same strings, e.g. g instead of grams. All other words, like _sugar_ in _"23 grams sugar"_ also needs to be removed from the expected test results. I will fix this when we have a way forward for the above. DS.

Reply:
Excellent observation! In this case, strings.Cut won't cut it so your original idea is better. 

Another possible way to handle the above is to first extract the number from the string using `regex.Digit.FindString` then append the appropriate abbreviated unit. This way we avoid trimming all possible variations and typos.

Reply:
I woke up this morning with the exact same idea. 👍Don't know why I did not think of extracting only the numbers yesterday.
Thank you for mentioning `regex.Digit.FindString`. That will definitely save me a bit of time as this is the first time I code anything in Go.

I also wanted to check with you the purpose behind line 3 in the below code that extracts the nutrition.servingSize. Why do we have the conditions `len(xs) == 2 && len(xs[1]) < 4`?
```go
if val := extensions.ConvertToString(x["servingSize"]); val != "" {
	xs := strings.Split(val, " ")
	if len(xs) == 2 && len(xs[1]) < 4 {
		n.Servings = val
	} else {
		for _, s := range xs {
			_, err := strconv.Atoi(s)
			if err == nil {
				n.Servings = s
				break
			}
		}
	}
}
```

Reply:
@reaper47 I am currently going through and updating all test cases to accomodate for the new `ensureNutritionUnitForString` function.

I got one question/suggestion, wouldn't it be even better if we only store the number values instead of adding the units at all?
Then we could simply add the unit in the frontend when viewing a recipe instead.

Reply:
Excellent, that should do it! Let's ignore cases where the unit is incorrect as they happen seldom. We could always handle those cases in the future if ever we notice they happen more frequently than previously thought.

Regarding the conditional block above, the if condition handles cases where `servingSize` contains a unit, e.g. "20 g", "2 TBS", "serves 6", "0.5 cup". The else handles the rest, e.g. "1 serving", "1 serving (based on 4)", "1". The use case for this is to determine whether the nutrition facts should be displayed per serving or per 100g.  

We could leave it as is for the time being because you opened an interesting issue where the user can toggle between `per serving` and `per 100g`. This block could then be rewritten to the following to differentiate between servings based on a quantity (with a unit) and a number of portions (without a unit).
```golang
if val, ok := x["servingSize"].(string); ok {
	v := regex.Unit.FindString(strings.ToLower(val))
	if v != "" {
		n.Servings = val
	} else {
		n.Servings = regex.Digit.FindString(val)
	}
}
```

Solely storing the number is actually a great idea because we'll save a bit of bytes and repetition in the database and the units are kind of standard. The type of each NutritionSchema.<nutrition> should remain `string` because it would be too much of a hassle to alter the [nutrition table](https://github.com/reaper47/recipya/blob/main/internal/services/migrations/20230609214207_add_recipes.sql#L57-L71).

Reply:
I agree. 👍
I will change the new function to only keep the digits and not add any units as suffix. With that change we may not even need the `ensureNutritionUnitForString` function, but I'll keep it so it is easy to implement handling of any edge cases in the future.

Do you know where changes are needed to display the units in the frontend, since we now will only be storing the digits?

Regarding the servingSize, I still am unclear regarding that condition. E.g. why do we check the length of less than 4 characters at `len(xs[1]) < 4`?
I assume the 4 comes from experimentation and that you have determined that in most cases all strings you want to ignore contains max 3 characters. Is that a correct assumption?
If so, then recipes that produces for instant "1 bar" (like a candy bar, protein bar etc.) will get incorrect servingSize I believe.
No judgement, as I said, just wanted to understand the reason behind it as I was perplexed when reading that condition. 🙂
Either way, your suggested change above will solve this I think and we'll skip this for now and handle it via #385.

Reply:
Perfect! This is where the nutrition table is rendered in the frontend: https://github.com/reaper47/recipya/blob/main/web/components/recipes.templ#L2126. 

The print view of the nutrition is rendered [here](https://github.com/reaper47/recipya/blob/main/web/components/recipes.templ#L2236). The [Format function](https://github.com/reaper47/recipya/blob/main/internal/models/recipe.go#L506) should also be checked.

I am also perplexed as to how I came up with that, but I do believe it was indeed from experimentation.

Reply:
Thanks. I have tested and know where to change the frontend now.

A question regarding people's existing recipes, when this change has been merged into the main branch.
Since existing recipes will follow the old scraper, meaning that they might say _g_, _gram_, _grams_ etc., how easy would it be to handle this as some type of database migration so we make sure that we only store the digits both for new and already existing recipes? If we go this route, the frontend should probably not allow the end-user to type in custom strings in the nutritional fields, but only digits.

If the above is time consuming I guess the alternative is to add some logic when handling the strings. This might instead become cumbersome to manage in the future when translations as of #260 is implemented (depending on how that will be handled on the backend)

Reply:
The reason why you cannot use `EnsureNutritionUnitForString` is because the `nutritionProperty` parameter is missing in the code above. In Go, unlike JavaScript and C#, basic function parameters cannot be omitted nor can they have default values. You will sometimes see `func (a, b string, opts ...string)`, where `a` and `b` are mandatory, whereas `opts` can be omitted. The latter is represented as a slice of strings in the function's body.

For now, we can do the digit-extraction work in the frontend because doing otherwise would involve quite some work. This will indeed remove custom strings users have typed, which is fine because I'm not sure many people calculate/type the nutrition values themselves. The third option is great because we could then regenerate the data in a standard way.

The database migration should be easy using Go and SQL statements. We could have a function that does the migration on program startup. Writing a database migration might be more difficult. Ideally, the columns in the `nutrition` table should be altered to `INTEGER`, the add/edit table accepts only numerical values and displays standard units next to the input boxes, and the frontend only displays the number from the database without any additional work.

Reply:
Yeah, that was my thought as well as I noticed the nutritionProperty was missing after I posted my last reply, but that unfortunately did not make a difference.
And I do not get the intellisense to suggest the function even when typing models.Ensure... either, which works just fine when trying to reference the same function from other files, so there must be something else I am missing. 🤔

I agree, I will implement the third option. 👍

Reply:
In this case, it might be because you need to update `templ`. You can run `go get -u github.com/a-h/templ` to upgrade it. It will update the go.mod and go.sum files. You might also need to update your binary by running `go install github.com/a-h/templ/cmd/templ@latest`. As for intellisense, I don't know whether it works in vscode with the [templ extension](https://marketplace.visualstudio.com/items?itemName=a-h.templ). I do know that it doesn't work in Goland linux. 

I managed to build your branch with `task build` after writing and updating templ:
```html
<tr>
	<td>Fiber:</td>
	<td>
		if data.Recipe.Nutrition.Fiber == "" {
		    -
		} else {
		    { models.EnsureNutritionUnitForString(extensions.ConvertToString(data.Recipe.Nutrition.Fiber), "fiberContent") }
		}
	</td>
</tr>
```

Reply:
Thank you for troubleshooting it on your end.
I can confirm that it is indeed the Go intellisense in VS Code that either needs updating or simply does not work in templ files, as directly when I saw your reply just now, I simply ran `task build-debug` and I can see my function is called perfectly fine and rendered in the UI afterwards.
Whilst the above works fine, I still get the same intellisense error:
![image](https://github.com/reaper47/recipya/assets/27766499/e9b9bfb3-8671-48c0-9709-cef92f7fa837)

Glad to know it wasn't me after all, but I really should have tried to rebuild and actually test it out yesterday before troubleshooting what was a non-existent issue. 🤦‍♂️

I will continue to implement this change and update all test cases to accommodate for the change.

Reply:
I noticed that the [Format function](https://github.com/reaper47/recipya/blob/main/internal/models/recipe.go#L506) is missing the sodium property, which I assume is a mistake. If there is a reason this was excluded from the print view, let me know.

It is also missing the trans fat and unsaturated fat properties, but that goes for both the UI and the print view as mentioned in issue #386.

Reply:
Glad to hear everything's fine and the problem was a false flag on the templ extension's side.

Nice catch! I did indeed miss the sodium property in the Format function and the two other fat properties in the nutrition table :/ They shouldn't have been excluded.

Reply:
I have added the _Unsaturated fat_ property, but I noticed that _Trans fat_ is missing from the nutrition type declaration so I did not want to touch that in this PR.

All test cases has been updated to accomodate for the digits only extraction of nutritional information and now passes OK.